presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java

```diff
@@ -68,74 +74,98 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
         }, value);
     }
 
-    public AggregationNode map(AggregationNode node, PlanNode source)
+    public <T extends PlanNode> T map(PlanNode planNode, PlanNodeIdAllocator idAllocator, RecursionStrategy recursionStrategy)
```

Comment:
* I really like the overall goal of having a general SymbolMapper facility, and using it in UnaliasSymbolReferences (or of course its iterative equivalent) as a special case.  That will also easily allow a general iterative rename-pushdown optimization.
* I don't think SymbolMapper needs to expose RecursionStrategy to its user.  Recursive uses can use generic function like this patch has, and the previous interface already worked for non-recursive uses.
* Could you explain how the SymbolMapper builder and class is a better idea than having a static function for each Node type that takes a "fromTo" map as an explicit argument?  The only thing I can come up with is that IntelliJ can prompt the user of the builder with the argument names "from" and "to", but maybe I'm missing something.